### ISSUE TITLE
RecordStatesAsync should return something when its finished

### ISSUE THREAD

**Issue by grobinson-grafana** [grafana/grafana#61811]
**What happened**:

The `RecordStatesAsync` method does not provide a way for the caller to know if the recording of states has completed.

```go
// Historian maintains an audit log of alert state history.
type Historian interface {
	// RecordStates writes a number of state transitions for a given rule to state history.
	RecordStatesAsync(ctx context.Context, rule *models.AlertRule, states []StateTransition)
}
```

This can be problematic when it comes to cancelation of contexts, such that the goroutines for individual contexts can be kept in memory much longer than is necessary.

For example, here the context will live for at least 5 seconds, even if the states are recorded in just 100ms. The context cannot be garbage collected until at least 5 seconds after it has been created.

```go
ctx, _ := context.WithTimeout(context.Background(), 5 * time.Second)
historian.RecordStatesAsync(ctx, rule, states)
```

To fix this I propose that `RecordStatesAsync` returns a channel which is closed on completion.

```go
// Historian maintains an audit log of alert state history.
type Historian interface {
	// RecordStates writes a number of state transitions for a given rule to state history.
	RecordStatesAsync(ctx context.Context, rule *models.AlertRule, states []StateTransition) <-chan struct{}
}
```

The caller can then do something like this:

```go
ctx, cancelFunc := context.WithTimeout(context.Background(), 5 * time.Second)
go func() {
    defer cancelFunc()
    <-historian.RecordStatesAsync(ctx, rule, states)
}()
```

**Comment by alexweav** [comment]
I don't think we have ever created new contexts specifically for recording states, we always just inherited the state manager's context, even in the old code before I refactored it this was the case. Does this actually happen in the context that it's used?

We've always spawned that goroutine, and it's never had a timeout. We just let it run in the background. Maybe we _should_, but if so, that's something that all of our background operations should do.

I don't think the context that we're creating in the provided examples exists in the system right now to begin with, so there isn't currently anything to collect, unless I'm missing something?

**Comment by grobinson-grafana** [comment]
Yeah, at present we use `grafanaCtx` which I understand doesn't have a timeout, but should be canceled when the rule is deleted. In Alex's pull request we need to call `RecordStatesAsync` when this happens, but we cannot use `grafanaCtx` because it is canceled. We also don't want to use `context.Background()` because there is no upper bound on the lifetime of the goroutine.

> We've always spawned that goroutine, and it's never had a timeout. 

It's not about timing out, but the about the goroutine never terminating because there is no timeout. Do we know that `RecordStatesAsync` always completes and never hangs?

**Comment by alexweav** [comment]
I think there is another task we'll need for this as well:

`RecordStates` takes pointers to rules and also states (inside StateTransition). The current code is extremely careful to copy all values it needs out of these structs, in the synchronous parts, before starting any goroutines. Once we're in a concurrent context these objects can change out from underneath us, either by requests updating rules, or by new evaluations changing the states under the pointers - we can end up recording the wrong data.

If we are going to change the concurrency to be fully at the call site, we'll need to solve this first, or it introduces a bug here - probably, `RecordStates`, should not take pointers to anything, and just copy structs.
